clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -5348,18 +5348,8 @@ Value *ScalarExprEmitter::VisitVAArgExpr(VAArgExpr *VE) {
     return llvm::UndefValue::get(ArgTy);
   }
 
-  // FIXME Volatility.
-  llvm::Value *Val = Builder.CreateLoad(ArgPtr);
-
-  // If EmitVAArg promoted the type, we must truncate it.
-  if (ArgTy != Val->getType()) {
-    if (ArgTy->isPointerTy() && !Val->getType()->isPointerTy())
-      Val = Builder.CreateIntToPtr(Val, ArgTy);
-    else
-      Val = Builder.CreateTrunc(Val, ArgTy);
-  }
-
-  return Val;
+  return CGF.EmitLoadOfScalar(ArgPtr, Ty.isVolatileQualified(), Ty,
+                              VE->getExprLoc());
```

Comment:
> That seems like it'd be wrong on big-endian targets if the variadic argument was promoted.

I don't think we actually depend on this behavior anymore; we either adjust the pointer to point to the correct bytes, or store the value in a temporary.

Reply:
Should we just make `EmitVAArg` return an `RValue`?  Sema appears to unconditionally consider it a pr-value, so while CodeGen has an l-value path for it, that's probably dead.

Reply:
> Should we just make EmitVAArg return an RValue?

You mean, using the same signature as EmitAnyExpr?  That should work, sure.

> while CodeGen has an l-value path for it, that's probably dead.

I think we end up using EmitVAArgExprLValue in C, for something like `va_arg(x, struct X).x`.  But that's not a real lvalue; it's just a temporary.

Reply:
Do you mean I should apply `RValue::get()` in `CodeGenFunction::EmitVAArg`? And all the callers of `EmitVAAArg` therefore should do `getScalarVal()` or similar to get a `llvm::value`?
Or is it still better to do in another PR?
If I revert the current change with `EmitLoadOfScalar` there should be a manipulation for _BitInt like it was in first version of the patch and received this comment - https://github.com/llvm/llvm-project/pull/91364#discussion_r1592864982 .

Reply:
> And all the callers of EmitVAAArg therefore should do getScalarVal() or similar to get a llvm::value

Yes, that's the idea.  Should significantly simplify the code for some targets.

This should be a dependent PR; we merge that first, then this PR doesn't need to touch the varargs code.

Reply:
I posted https://github.com/llvm/llvm-project/pull/94635 .

I somehow had to load the value anyway, since inner `EmitVAArg` that seems to be target-dependent returns an Address as well.